pkg/trigger/discovery_handler.go

```diff
@@ -185,10 +185,35 @@ func (mt *MigrationTrigger) processDiscoveryResource(r metav1.APIResource) {
 		utilruntime.HandleError(getErr)
 		return
 	}
-
-	stale := (getErr == nil && mt.staleStorageState(ss))
-	storageVersionChanged := (getErr == nil && ss.Status.CurrentStorageVersionHash != r.StorageVersionHash)
-	notFound := (getErr != nil && errors.IsNotFound(getErr))
+	stale := getErr == nil && mt.staleStorageState(ss)
+	storageVersionChanged := getErr == nil && ss.Status.CurrentStorageVersionHash != r.StorageVersionHash
+	notFound := getErr != nil && errors.IsNotFound(getErr)
+
+	var relaunchMigration bool
+	if stale || notFound || storageVersionChanged {
+		relaunchMigration = true
+	} else {
+		// get the corresponding StorageVersionMigration resource
+		migrations, err := mt.migrationInformer.GetIndexer().ByIndex(controller.ResourceIndex, controller.ToIndex(toGroupResource(r)))
+		if err != nil {
+			utilruntime.HandleError(getErr)
+			return
+		}
+		switch {
+		case len(migrations) == 0:
+			// corresponding StorageVersionMigration resource missing
+			relaunchMigration = true
+		case len(ss.Status.PersistedStorageVersionHashes) == 0:
+			// this should not happen
+			relaunchMigration = true
+		case ss.Status.PersistedStorageVersionHashes[0] == migrationv1alpha1.Unknown:
+			migration := migrations[0].(*migrationv1alpha1.StorageVersionMigration)
+			if controller.HasCondition(migration, migrationv1alpha1.MigrationFailed) {
```

Comment:
Can you describe what leads to this situation? The kubemigrator does not give up easily.

Reply:
See https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kube-storage-version-migrator/47/pull-kube-storage-version-migrator-disruptive/1224770718261055490/build-log.txt

```
Feb  4 19:44:47.629: INFO: resource {migration.k8s.io storagestates} has persisted hashes [Unknown], and current hash 7abAo0yHdNM=
Feb  4 19:44:47.629: INFO: timed out waiting for the condition
```

```yaml
- apiVersion: migration.k8s.io/v1alpha1
  kind: StorageVersionMigration
  metadata:
    creationTimestamp: "2020-02-04T19:26:54Z"
    generateName: storagestates.migration.k8s.io-
    generation: 1
    name: storagestates.migration.k8s.io-kk589
    resourceVersion: "3574"
    selfLink: /apis/migration.k8s.io/v1alpha1/storageversionmigrations/storagestates.migration.k8s.io-kk589
    uid: 7dc9f088-aeac-49f8-bbf7-a31a25b55fdb
  spec:
    resource:
      group: migration.k8s.io
      resource: storagestates
      version: v1alpha1
  status:
    conditions:
    - lastUpdateTime: "2020-02-04T19:30:49Z"
      message: '[Put https://10.0.0.1:443/apis/migration.k8s.io/v1alpha1/storagestates/jobs.batch:
        unexpected EOF, Put https://10.0.0.1:443/apis/migration.k8s.io/v1alpha1/storagestates/statefulsets.apps:
        unexpected EOF]'
      status: "True"
      type: Failed
```
```yaml
- apiVersion: migration.k8s.io/v1alpha1
  kind: StorageState
  metadata:
    creationTimestamp: "2020-02-04T19:26:54Z"
    generation: 1
    name: storagestates.migration.k8s.io
    resourceVersion: "5683"
    selfLink: /apis/migration.k8s.io/v1alpha1/storagestates/storagestates.migration.k8s.io
    uid: 4301aff1-bb56-4a9f-8562-cac09ce5031a
  spec:
    resource:
      group: migration.k8s.io
      resource: storagestates
  status:
    currentStorageVersionHash: 7abAo0yHdNM=
    lastHeartbeatTime: "2020-02-04T19:44:11Z"
    persistedStorageVersionHashes:
    - Unknown
```

Reply:
Got it. I think the "Put https://10.0.0.1:443/apis/migration.k8s.io/v1alpha1/storagestates/jobs.batch: unexpected EOF" message means that the apiserver restarts in the middle of handling the PUT request from the migrator/core. We should let the migrator/core retry in that case, to avoid re-migrating the entire list of Batches.

Apart from that, I agree that we should launch a migration if ss.Status.PersistedSotrageVersionHashes!=ss.Status.CurrentStoragteVersionHash && (there is no pending or running storageMigration). Instead of a switch block nested in the else clause, I would make another variable to capture this state, something like this:

```golang
func (mt *MigrationTrigger) processDiscoveryResource(r metav1.APIResource) {
        ...
	stale := (getErr == nil && mt.staleStorageState(ss))
        notFound := (getErr != nil && errors.IsNotFound(getErr))
        // migrated checks if currentStorageVersionHash==persistedStorageVersionHashes, 
        needMigration := !migrated(ss) && !pendingOrRuningMigrations(ss)
        // If storage version has changed, we need to restart any running storageMigration.
	storageVersionChanged := (getErr == nil && ss.Status.CurrentStorageVersionHash != r.StorageVersionHash)
	if stale {
            ....
        }
        if stale || needMigration || storageVersionChanged || notFound {
             relaunch...
        }
}

Reply:
@caesarxuchao Done. PTAL.